### DISCUSSION TITLE
Service adguardhome writes persistent data to SD card

### DISCUSSION THREAD

**Discussion by cmonty14** [MichaIng/DietPi#7048] [Q&A]
Hello,
I deployed software 
- AdGuard Home (ID: 126)
- Unbound (ID: 182)

successfully on my (old) RPi B (armv6l).

Thanks to your pre-configured deployment this runs like charm.

In my understanding _AdGuard Home_ is writing persistent data to directory `/mnt/dietpi_userdata/adguardhome/data`.
If this is true, this undermines your approach with _dietpi-ramlog_ to save the lifetime of the SD card.

Could you provide a solution that creates a tmpfs for `/mnt/dietpi_userdata/adguardhome/data` and persists this data with a cronjob?
Or any other solution that saves the lifetime of the SD card.

Regards
Thomas

**Comment by Joulinar** [comment]
> this undermines your approach with dietpi-ramlog 

Well, `dietpi-ramlog` does not aim to prevent every single write operation or to keep application data in memory. The tool is about exactly what the name says, keeping system logs in memory but not application data.

Theoretically, you could ask the AGH developers if they have a similar function to PiHole. With PiHole you can set how often the data is written to the disc. The default setting is every 10 minutes. I don't know if AGH offers something similar.

**Comment by th2j** [comment]
Hello

Here is a solution to put the adguard data folder in RAM :

**1**
**sudo nano /etc/systemd/system/startupadguard.service**

paste :
```
[Unit]
Description=startupadguard
Wants=
After=network-online.target

[Service]
Type=simple
User=root
Group=root
ExecStart=/bin/bash /usr/local/bin/startupadguard.sh

[Install]
WantedBy=multi-user.target
```

**sudo nano /usr/local/bin/startupadguard.sh**

paste :
```
#!/bin/bash
exec &>/home/dietpi/startupadguard.log

sleep 6

systemctl stop adguardhome
while [ $(pgrep AdGuardHome) > 0 ]; do sleep 1; done
sleep 2

umount /media/adguard/data
mount -t tmpfs -o size=256M,uid=adguardhome,gid=adguardhome,mode=0700 tmpfs /media/adguard/data

rm /mnt/dietpi_userdata/adguardhome/data
rm -r /mnt/dietpi_userdata/adguardhome/data
ln -s /media/adguard/data /mnt/dietpi_userdata/adguardhome/data
cp -p /mnt/dietpi_userdata/adguardhome/data-backup/* /media/adguard/data/

systemctl start adguardhome

exit 0
```

**sudo chmod +x /usr/local/bin/startupadguard.sh
**sudo systemctl daemon-reload
sudo systemctl enable startupadguard****

then disable adguard autostart in dietpi-services

**2**
create a cron job to backup data

**sudo nano /etc/cron.d/cron-dietpi**

paste :
`15 6 * * * root cp -p /media/adguard/data/* /mnt/dietpi_userdata/adguardhome/data-backup/`

**sudo service cron reload**

**3**
****sudo mkdir /media/adguard
sudo mkdir /media/adguard/data
sudo chown adguardhome:adguardhome /media/adguard
sudo chown adguardhome:adguardhome /media/adguard/data
sudo cp -pr /mnt/dietpi_userdata/adguardhome/data /mnt/dietpi_userdata/adguardhome/data-backup** 
sudo rm -r /mnt/dietpi_userdata/adguardhome/data-backup/filters**

then reboot, it should be fine
it takes 256mb in RAM so /mnt/dietpi_userdata/adguardhome/data must not use more space

**Reply by MichaIng** [reply]
Instead of creating a dedicated service, just attach your script to the original service.

Do once to copy data to backup dir and establish the symlink
```sh
sudo systemctl stop adguardhome
sudo -u adguardhome mkdir -p /mnt/dietpi_userdata/adguardhome/data-backup
sudo cp -a /mnt/dietpi_userdata/adguardhome/data/* /mnt/dietpi_userdata/adguardhome/data-backup/
sudo rm -R /mnt/dietpi_userdata/adguardhome/data
sudo mkdir -p /media/adguard/data
sudo -u adguardhome ln -s /media/adguard/data /mnt/dietpi_userdata/adguardhome/data
```
Create and invoke pre-start and post-stop scripts to mount/unmount the tmpfs and copy back/forth data. That way, also on service stop and restart, data is synced to disk.
```sh
sudo mkdir -p /etc/systemd/system/adguardhome.service.d
cat << '_EOF_' | sudo tee /etc/systemd/system/adguardhome.service.d/data-to-ram.conf
[Service]
ExecStartPre=/usr/local/bin/adguard-to-ram.sh
ExecStopPost=/usr/local/bin/adguard-to-disk.sh
_EOF_
cat << '_EOF_' | sudo tee /usr/local/bin/adguard-to-ram.sh
#!/bin/dash
set -e
findmnt /media/adguard/data > /dev/null || mount -t tmpfs -o size=256M,uid=adguardhome,gid=adguardhome,mode=0700 tmpfs /media/adguard/data
cp -a /mnt/dietpi_userdata/adguardhome/data-backup/* /media/adguard/data/
_EOF_
cat << '_EOF_' | sudo tee /usr/local/bin/adguard-to-disk.sh
#!/bin/dash
set -e
[ ! -d '/media/adguard/data' ] || cp -a /media/adguard/data/* /mnt/dietpi_userdata/adguardhome/data-backup/
! findmnt /media/adguard/data > /dev/null || umount /media/adguard/data
_EOF_
sudo chmod +x /usr/local/bin/adguard-to-ram.sh /usr/local/bin/adguard-to-disk.sh
systemctl daemon-reload
```
The cron job can be best be added to the `adguardhome` crontab as well:
```sh
echo '15 6 * * * cp -a /media/adguard/data/* /mnt/dietpi_userdata/adguardhome/data-backup/' | sudo crontab -u adguardhome -
```
For others: above cron job will execute once a day at 06:15 am. To run it once an hour, use e.g. `15 * * * *` to run it every hour at x:15. Or `15 */3 * * *` for every 3 hours at x:15 etc. All data after last sync is lost on system craft, so it depends on how much you trust in stability of your device (including e.g. electricity outages), and how important the query logs and such are for you.

@th2j 
Why do you remove `/mnt/dietpi_userdata/adguardhome/data-backup/filters`?

**Reply by th2j** [reply]
> @th2j Why do you remove `/mnt/dietpi_userdata/adguardhome/data-backup/filters`?

You can keep it, but as it is downloaded at adguardhome startup, i don't really need it
This way i limit writing on disk

Thanks
